src/keys.c

```diff
@@ -3088,8 +3131,28 @@ static int MakeSslMasterSecret(WOLFSSL* ssl)
             idx = pmsSz;  /* preSz */
             XMEMCPY(md5Input + idx, shaOutput, SHA_DIGEST_SIZE);
             idx += SHA_DIGEST_SIZE;
-            wc_Md5Update(md5, md5Input, idx);
-            wc_Md5Final(md5, &ssl->arrays->masterSecret[i * MD5_DIGEST_SIZE]);
+            ret = wc_Md5Update(md5, md5Input, idx);
+            if (ret != 0) {
+            #ifdef WOLFSSL_SMALL_STACK
+                XFREE(shaOutput, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+                XFREE(md5Input,  NULL, DYNAMIC_TYPE_TMP_BUFFER);
+                XFREE(shaInput,  NULL, DYNAMIC_TYPE_TMP_BUFFER);
+                XFREE(md5,       NULL, DYNAMIC_TYPE_TMP_BUFFER);
+                XFREE(sha,       NULL, DYNAMIC_TYPE_TMP_BUFFER);
+            #endif
+                return ret;
+            }
+            ret = wc_Md5Final(md5, &ssl->arrays->masterSecret[i * MD5_DIGEST_SIZE]);
+            if (ret != 0) {
+            #ifdef WOLFSSL_SMALL_STACK
+                XFREE(shaOutput, NULL, DYNAMIC_TYPE_TMP_BUFFER);
```

Comment:
Too much code is added here.  Notice that other failures drop down to bottom to free in one place.